src/algorithms/policy_gradient/vpg.jl

```diff
@@ -0,0 +1,156 @@
+"""
+Base.@kwdef mutable struct VPGPolicy{
+    A<:NeuralNetworkApproximator,
+    B<:Union{NeuralNetworkApproximator,Nothing},
+    R<:AbstractRNG,
+} <: AbstractPolicy
+    approximator::A
+    baseline::B = nothing
+    dist::Any = Categorical
+    γ::Float32 = 0.99f0 # discount factor
+    α_θ = 1.0f0 # step size of policy
+    α_w = 1.0f0 # step size of baseline
+    batch_size::Int = 1024
+    rng::R = Random.GLOBAL_RNG
+    loss::Float32 = 0.0f0
+    baseline_loss::Float32 = 0.0f0
+end
+
+function (agent::Agent{<:VPGPolicy,<:AbstractTrajectory})(::Training{PreActStage}, env)
+    action = agent.policy(env)
+    push!(agent.trajectory; state = get_state(env), action = action)
+    if ActionStyle(env) === FULL_ACTION_SET
+        push!(agent.trajectory; legal_actions_mask = get_legal_actions_mask(env))
+    end
+    update!(agent.policy, agent.trajectory, get_actions(env))
+    action
+end
+
+function (agent::Agent{<:VPGPolicy,<:AbstractTrajectory})(::Training{PostEpisodeStage}, env)
+    action = agent.policy(env)
+    push!(agent.trajectory; state = get_state(env), action = action)
+    if ActionStyle(env) === FULL_ACTION_SET
+        push!(agent.trajectory; legal_actions_mask = get_legal_actions_mask(env))
+    end
+    update!(agent.policy, agent.trajectory, get_actions(env))
+    action
+end
+
+function (π::VPGPolicy)(env)
+    to_dev(x) = send_to_device(device(π.approximator), x)
+
+    logits = env |> get_state |> to_dev |> π.approximator
+    π(logits, get_actions(env))
+end
+
+function (π::VPGPolicy)(env::MultiThreadEnv)
+    error("not implemented")
+    # TODO: can PG support multi env? PG only get updated at the end of an episode.
+end
+
+function (π::VPGPolicy)(logits, actions::DiscreteSpace)
+    dist = logits |> softmax |> π.dist
+    action = actions[rand(π.rng, dist)]
+end
+
+"""
+See
+* [Diagonal Gaussian Policies](https://spinningup.openai.com/en/latest/spinningup/rl_intro.html#stochastic-policies
+* [Clipped Action Policy Gradient](https://arxiv.org/pdf/1802.07564.pdf)
+"""
+function (π::VPGPolicy)(logits, ::ContinuousSpace)
+    dist = π.dist.(logits...)
+    action = rand.(π.rng, dist)[1]
+end
+
+function (π::VPGPolicy)(logits, actions::MultiDiscreteSpace)
+    error("not implemented")
+    # TODO
+end
+
+function (π::VPGPolicy)(logits, actions::MultiContinuousSpace)
+    error("not implemented")
+    # TODO
+end
+
+function RLBase.update!(π::VPGPolicy, traj::ElasticCompactSARTSATrajectory)
+    error("not supported")
+end
+
+function RLBase.update!(π::VPGPolicy, traj::ElasticCompactSARTSATrajectory, action_space)
+    (length(traj[:terminal]) > 0 && traj[:terminal][end]) || return
+
+    model = π.approximator
+    to_dev(x) = send_to_device(device(model), x)
+
+    states = traj[:state]
+    actions = traj[:action]
+    gains = traj[:reward] |> x -> discount_rewards(x, π.γ)
+
+    for idx in Iterators.partition(shuffle(1:length(traj[:terminal])), π.batch_size)
+        S = states[:, idx] |> to_dev
```

Comment:
Use `view` here to avoid unnecessary memory allocation. To support `states` of higher dimension, you're suggested to use the `select_last_dim` function provided in [`RLCore`](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/a055fdad9324868d97fea7f7a31ba2b7a091051e/test/utils/base.jl#L14)

Reply:
about `view`.
I found out this doc: https://docs.julialang.org/en/v1/manual/performance-tips/#Copying-data-is-not-always-bad

and did a benchmark:
`
julia> N=10000; X=randn(N); A=randn(N); inds = shuffle(1:N);
julia> @time sum(X.*A); @time sum(X[inds].*A[inds]); @time @views sum(X[inds].*A[inds]);
  0.000026 seconds (5 allocations: 78.281 KiB)
  0.000093 seconds (9 allocations: 234.688 KiB)
  0.000052 seconds (7 allocations: 78.469 KiB)
`

@views is faster than using slice directly.

I think the example in the julia doc did not count the time cost of memory allocation.

Reply:
Thanks for the link.

[Copying-data-is-not-always-bad](https://docs.julialang.org/en/v1/manual/performance-tips/#Copying-data-is-not-always-bad) is correct. Usually, the algorithm will run on GPU and the data will be transformed into `Array` first:

https://github.com/JuliaGPU/CUDA.jl/blob/f31cbe22b4baba872a48bcb48e9f60e712f653fc/src/array.jl#L206

So there's no difference whether we use view or not in this case.

In CPU only environments I guess we can still add a *copy* layer to address the performance issue?

Reply:
I benchmarked again, with larger 2d arrays.
using array slicing is much **faster** than `@views`.

```
julia> a=randn(1000,2000); b=randn(2000,1000);
julia> ind1 = shuffle(1:1000); ind2 = shuffle(1:2000);

julia> @time a*b;
  0.014062 seconds (2 allocations: 7.629 MiB)

julia> @time a[ind1,ind2]*b[ind2,ind1]; ## array slicing. fast, cost more memory
  0.027007 seconds (6 allocations: 38.147 MiB, 7.27% gc time)

julia> @time @views a[ind1,ind2]*b[ind2,ind1]; ## views. slow. does not require extra memory
  1.490200 seconds (14 allocations: 7.630 MiB)
```

I think in RL, the data size is very small usually, so there is no difference between slicing and `@views`.

For performance optimization, I think we should use a profiler to analyze the time cost of the code first.

what is a "copy layer"?
Can we simply use array slicing, or data |> Array?

```
julia> @time @views Array(a[ind1,ind2])*Array(b[ind2,ind1]);
  0.032226 seconds (12 allocations: 38.147 MiB)
```
Convert view to Array is still slower than array slicing, but much faster than using views.

Reply:
> what is a "copy layer"?
> Can we simply use array slicing, or data |> Array?

Yes. That's what I mean here.

Reply:
I think we can split this into another issue.

1. We stick with the `view` and users need to manually add an extra layer in `approximator` when working with CPU only . Just like

```
Chain(
    x -> convert(Array, x),
    # some other layers
)
```

2. Change the implementation of [`send_to_host`](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/a94544cfceb4a6d527895414ead2744f66f23949/src/utils/device.jl#L12)

So we convert the `SubArray` from `view` to `Array` before feeding it into the approximator.

I prefer the latter.

Reply:
In any case, `states[:, idx]` still needs to be updated to handle states of higher dimensions.

Reply:
Created an issue here: https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/issues/130